src/backend/distributed/utils/colocation_utils.c

```diff
@@ -0,0 +1,206 @@
+ *
+ *-------------------------------------------------------------------------
+ */
+
+#include "postgres.h"
+
+#include "access/genam.h"
+#include "access/heapam.h"
+#include "access/htup_details.h"
+#include "distributed/colocation_utils.h"
+#include "distributed/metadata_cache.h"
+#include "distributed/multi_logical_planner.h"
+#include "utils/rel.h"
+
+
+static List * DistTableOidList(void);
+
+
+/*
+ * TableColocationId function returns co-location id of given table. This function errors
+ * out if given table is not distributed.
+ */
+int
+TableColocationId(Oid distributedTableId)
+{
+	DistTableCacheEntry *cacheEntry = DistributedTableCacheEntry(distributedTableId);
+
+	return cacheEntry->colocationId;
+}
+
+
+/*
+ * TablesColocated function checks whether given two tables are co-located and
+ * returns true if they are co-located. This function errors
+ * out if given tables is not distributed.
+ */
+bool
+TablesColocated(Oid leftDistributedTableId, Oid rightDistributedTableId)
+{
+	int leftColocationId = INVALID_COLOCATION_ID;
+	int rightColocationId = INVALID_COLOCATION_ID;
+
+	if (leftDistributedTableId == rightDistributedTableId)
+	{
+		return true;
+	}
+
+	leftColocationId = TableColocationId(leftDistributedTableId);
+	rightColocationId = TableColocationId(rightDistributedTableId);
+	if (leftColocationId == INVALID_COLOCATION_ID ||
+		rightColocationId == INVALID_COLOCATION_ID)
+	{
+		return false;
+	}
+
+	return leftColocationId == rightColocationId;
+}
+
+
+/*
+ * ShardsColocated function checks whether given two shards are co-located and
+ * returns true if they are co-located. Two shards are co-located only if;
+ * - Tables containing the shards are co-located.
+ * - Min/Max values of the shards are same.
+ */
+bool
+ShardsColocated(ShardInterval *leftShardInterval, ShardInterval *rightShardInterval)
+{
+	bool tablesColocated = TablesColocated(leftShardInterval->relationId,
+										   rightShardInterval->relationId);
+
+	if (tablesColocated)
+	{
+		/*
+		 * Having co-located tables implies that tables are partitioned by hash partition
+		 * therefore it is safe to use DatumGetInt32 here.
+		 */
+		int leftShardMinValue = DatumGetInt32(leftShardInterval->minValue);
+		int leftShardMaxValue = DatumGetInt32(leftShardInterval->maxValue);
+		int rightShardMinValue = DatumGetInt32(rightShardInterval->minValue);
+		int rightShardMaxValue = DatumGetInt32(rightShardInterval->maxValue);
+
+		bool minValuesEqual = leftShardMinValue == rightShardMinValue;
+		bool maxValuesEqual = leftShardMaxValue == rightShardMaxValue;
+
+		return minValuesEqual && maxValuesEqual;
+	}
+
+	return false;
+}
+
+
+/*
+ * ColocatedTableList function returns list of relation ids which are co-located
+ * with given table.
+ */
+List *
+ColocatedTableList(Oid distributedTableId)
+{
+	List *colocatedTableList = NIL;
+	List *distributedTableList = DistTableOidList();
+	ListCell *distributedTableCell = NULL;
+
+	foreach(distributedTableCell, distributedTableList)
```

Comment:
A clarification, when you said index, is that index on pg_dist_partition and utilize it via `systable_beginscan`?